Fix connection drop is not detected

kidroca opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->
Update NetInfo usage to test for internet connectivity rather than active network connection
Desktop specific NetInfo shim is removed - the external issue in the comment has been resolved and offline state is detected in Desktop

Remains the same:
- When network is lost offline state is detected instantly (like turning off wifi or enabling airplane mode)

New:
- If internet is unavailable it it takes a bit of time to be detected - up to 12 seconds - when we have a request
 running for +7 seconds we trigger connection recheck that takes about 5 sec. (or less if it fails immediately) 
- When internet is resumed it may take up to 10 sec. to detect, depends on platform 

### Fixed Issues
<!---
Please replace GH_LINK with the link to the GitHub issue this Pull Request is fixing.
Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the issue; otherwise, the linking will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<number-of-the-issue>

Do NOT only link the issue number like this: $ #<number-of-the-issue>
--->
$ https://github.com/Expensify/App/issues/6059

### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

- [ ] Test network connection being correctly detected as offline
- [ ] Test network connection being correctly detected as online

#### Web / Desktop 
- Open dev tools 
- Set network state to offline from dev tools or turn of wifi 

#### iOS / Android 
- simulate poor network conditions (available in Android emulator)
- For iOS I think it's only possible on a physical device

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

Try to simulate very poor or no network conditions
- [ ] Test network connection being correctly detected as offline
- [ ] Test network connection being correctly detected as online

#### Web / Desktop 
- Use on a tethered connection
- Use with wifi disabled (cable unplugged)
- Connected to a connection with no internet (e.g. internet disabled from router)

#### iOS / Android 
- Try in airplane mode
- Try with mobile data disabled and no wifi
- Connection with no internet (e.g. internet disabled from router)

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

### Screenshots
<!-- Add screenshots for all platforms tested. Pull requests won't be merged unless the screenshots show the app was tested on all platforms.-->

#### Web
⚠️ It was only possible to test turning network off because of the web proxy on DEV. When deployed to staging it should work like in the Desktop video
<!-- Insert screenshots of your changes on the web platform-->
https://user-images.githubusercontent.com/12156624/154784651-00265a68-6800-4054-98b3-e0e1337b5cce.mp4

#### Mobile Web
<!-- Insert screenshots of your changes on the web platform (from a mobile browser)-->
https://user-images.githubusercontent.com/12156624/154778978-f0398e1d-86e8-490c-9091-aaeb35b06844.mp4

#### Desktop
<!-- Insert screenshots of your changes on the desktop platform-->

https://user-images.githubusercontent.com/12156624/154789103-daa03231-d5cf-4ef7-ab06-d6af259b6839.mp4

#### iOS
<!-- Insert screenshots of your changes on the iOS platform-->
https://user-images.githubusercontent.com/12156624/154789414-055aac0a-cb92-4795-be0f-cb35e76416a1.mp4

#### Android
<!-- Insert screenshots of your changes on the Android platform-->
https://user-images.githubusercontent.com/12156624/154778361-85b9e53a-5dad-4926-861a-d127621fcf66.mp4

kidroca committed:
Delete libs/NetInfo - no longer needed

libs/NetInfo served to provide shim for Electron due to an issue
The issue is now resolved and this is no longer needed

kidroca committed:
Update connectivity check with `isInternetReachable`

kidroca committed:
Update NetInfo to v8

Regarding the breaking change - we're not using SSID information - it doesn't affect us

v8 Includes the following fixes:
8.0.0 (2022-02-10)

BREAKING CHANGES
it's possible this is a breaking change, depending on your app use case. If you relied on iOS SSID information and met Apple's strict criteria for accessing SSID, you need to set the new config value shouldFetchWiFiSSID to true. If you set it to true and do not meet the criteria your app may crash due to a memory leak. All versions prior to 7.1.12 would attempt to fetch the information regardless of permission, leak memory, and possibly crash. This change avoids that crash.

Bug Fixes
ios: avoid memory leak from ssid APIs by adding explicit config (#560) (fbf7c15), closes #420

7.1.11 (2022-02-08)
Bug Fixes
windows: fix race condition in WiFi connection details feature (#568) (0cd8132)

7.1.10 (2022-02-07)
Bug Fixes
android: use registerDefaultNetworkCallback so toggling airplane mode works (#571) (e8af2de)

7.1.9 (2022-01-26)
Bug Fixes
android: count native listeners / correctly disable listener if count == 0 (#569) (5ae16f6)

7.1.8 (2022-01-25)
Bug Fixes
windows: refactor implementation to avoid crashes (#564) (cc4bfa3)

7.1.7 (2021-12-20)
Bug Fixes
android: populate network value during initial module startup (#553) (c05080f)

7.1.6 (2021-12-13)
Bug Fixes
android: avoid send event when has no listener (#548) (cad47d8)

7.1.5 (2021-12-09)
Bug Fixes
android: use method-local ref to instance var for multi-thread safety #549 (#550) (81bbc87)

7.1.4 (2021-12-07)
Bug Fixes
android: try async state fetch as stale state workaround (#547) (937cf48), closes #542

kidroca committed:
Update Podfile.lock for react-native-netinfo

kidroca committed:
Use api url for internet checks

This is primarily to help local development and desktop
By default on web (desktop) the url for checking is `/`
And for local testing you'll always hit it matching localhost:8080
For Desktop since the app is served from electron - requesting `/`
would respond with OK even if there's no internet

kidroca committed:
Only trigger offline state if we're certain there's no internet

kidroca committed:
Faster offline detection

parasharrajat commented on src/libs/NetworkConnection.js:
```diff
@@ -86,6 +100,16 @@ function onReconnect(callback) {
     reconnectionCallbacks.push(callback);
 }
 
+NetowrkLib.registerConnectionCheckCallback(_.throttle(() => {
+    if (!unsubscribeFromNetInfo) {
+        return;
+    }
+
+    // Reinitializing NetInfo would recheck state and trigger listeners
+    unsubscribeFromNetInfo();
+    subscribeToNetInfo();
+}), 10 * 1000, {trailing: false});
```
```suggestion
NetowrkLib.registerConnectionCheckCallback(_.throttle(() => {
    if (!unsubscribeFromNetInfo) {
        return;
    }

    // Reinitializing NetInfo would recheck state and trigger listeners
    unsubscribeFromNetInfo();
    subscribeToNetInfo();
}), 10000, {trailing: false});
```

parasharrajat commented on src/libs/Network.js:
```diff
@@ -233,6 +238,7 @@ function processNetworkRequestQueue() {
             .catch((error) => {
                 // When the request did not reach its destination add it back the queue to be retried
                 const shouldRetry = lodashGet(queuedRequest, 'data.shouldRetry');
+                recheckConnectivity();
```
```suggestion
		recheckConnectivity();
		
                // When the request did not reach its destination add it back the queue to be retried
                const shouldRetry = lodashGet(queuedRequest, 'data.shouldRetry');
```

parasharrajat commented on src/CONST/index.js:
```diff
@@ -306,6 +306,7 @@ const CONST = {
         },
         MAX_REQUEST_RETRIES: 10,
         PROCESS_REQUEST_DELAY_MS: 1000,
+        MAX_PENDING_TIME: 7 * 1000,
```
```suggestion
        MAX_PENDING_TIME: 7000,
```

kidroca commented on src/libs/NetworkConnection.js:
```diff
@@ -47,19 +49,31 @@ function setOfflineStatus(isCurrentlyOffline) {
  * `disconnected` event which takes about 10-15 seconds to emit.
  */
+function subscribeToNetInfo() {
+    NetInfo.configure({
+        // By default, for web (including Electron) NetInfo uses `/` for `reachabilityUrl`
+        // When App is served locally or from Electron this would respond with OK even with no internet
+        // Using API url ensures reachability is tested over internet
+        reachabilityUrl: CONFIG.EXPENSIFY.URL_API_ROOT,
+        reachabilityTest: response => Promise.resolve(response.status === 200),
+    });
```
> This should be one time thing

Calling `NetInfo.configure` is (re)checking current network state and we want to invoke it to force a recheck here  `registerConnectionCheckCallback` 

I'll add a comment about it, so that someone doesn't decide to move it out

This is the only way I found to force a state recheck, without this there can be up to a minute delay until you receive network state event that you're offline

parasharrajat replied:
Oh. Thanks for adding the comment.

kidroca reviewed (commented):
Can I ask what's the problem with defining duration as `10 * 1000` is it against code conventions?

kidroca committed:
Document `NetInfo.configure` serves to recheck network state

kidroca committed:
Rename MAX_PENDING_TIME to MAX_PENDING_TIME_MS

kidroca committed:
Add PR suggestion

kidroca committed:
Remove `trailing` from recheck

`trailing: false` causes duplicate calls
removing it makes just one recheck request (as intended)

kidroca committed:
Fix typo

kidroca commented:
Ready for review

parasharrajat commented:
> Can I ask what's the problem with defining duration as 10 * 1000

It is very clear that 10000 is 10 sec. It is just extra calculation. Why do you think 10 * 1000 is better?

kidroca commented:
> It is very clear that 10000 is 10 sec. It is just an extra calculation. Why do you think 10 * 1000 is better?

Maybe for you. I have hard time distinguishing thousands without some kind of visual separator:
- `10000` (This is looking like 100 000 to me)
- vs
- `10,000`
- `10 * 1000`

It's common to use this pattern even for seconds: https://github.com/react-native-netinfo/react-native-netinfo/blob/6b01fee211df160d0afce70c5955681c75b24a5a/src/internal/defaultConfiguration.web.ts#L3-L12

And we used as well: https://github.com/Expensify/App/search?q=expiration+%3D+1000+*+60

parasharrajat commented on src/libs/NetworkConnection.js:
```diff
@@ -47,19 +49,32 @@ function setOfflineStatus(isCurrentlyOffline) {
+    // whether a user has internet connectivity or not.
+    unsubscribeFromNetInfo = NetInfo.addEventListener((state) => {
+        Log.info('[NetworkConnection] NetInfo state', false, state);
+        setOfflineStatus(!state.isInternetReachable);
```
Here we set the offline/online status based on `isInternetReachable` prop. I didn't find many details about it but I think this flag is set based on the result of the HTTP request.  
So when the user is online it seems right to check the flag but when the user is offline it will take some time to be set. 
I think we should do this. 

if `state.isConnected`
then  offline status = !state.isInternetReachable
Else if  `!state.isConnected` 
then offline status = false. 

When a user is not connected then we don't have to wait for this flag `state.isInternetReachable`. If it behaves the same then It's OK.

kidroca replied:
`isInternetReachable` is not always set from the result of a request
When `state.isConnected` is `false` - `state.isInternetReachable` is also `false`

`isInternetReachable` is set from `true` to `false` when the check request fails or takes longer than 5sec.

parasharrajat reviewed (changes requested):
I don't mind 10 * 1000. I just thinking that we use 5000 ,3000 and not 5 * 1000 in other places of our app. 

it makes more sense to use when you have to show the different parts of calculations. 
e.g. 
Secs in day = 24 * 60 * 60

parasharrajat commented on src/libs/Network.js:
```diff
@@ -42,8 +43,12 @@ function processRequest(request) {
         : request.data;
 
+    // If request is still in processing after this time, we might be offline
+    const timerId = setTimeout(recheckConnectivity, CONST.NETWORK.MAX_PENDING_TIME_MS);
```
Ok, I see. Thanks for the explanation.

parasharrajat commented:
Testing now...

parasharrajat reviewed (changes requested):
the offline messages are shown for some time when we load the app. But when we are already online, there should be no offline message. May be related to https://github.com/Expensify/App/pull/7825#discussion_r812167915. 


https://user-images.githubusercontent.com/24370807/155397876-58cd6548-d6fe-4113-a810-699149e154f4.mp4

parasharrajat reviewed (changes requested):
issue 2: The offline message comes and goes many times. 
Steps: 
2. Connect your device to a mobile hotspot.
1. Open the web app. 
2. Let the sync complete. 
3. Turn off mobile data on a mobile device. 
4. Wait... 
5. Offline message comes and goes. 

Issue 3: If you do above till 4, you will notice that wait time is very high. Why? maybe because all the timers are cleared as all the requests are completed. Now when you go offline, It does not check internet reachability until a new request is sent. I think it is a Log request that triggers it after a big delay. 


https://user-images.githubusercontent.com/24370807/155400796-62454236-3d43-4ee2-8e1e-c7e65ca21afa.mp4

kidroca committed:
NetworkConnection only treat isInternetReachable === false as being offline

kidroca committed:
Replace throttle logic with lock flag

The recheck was getting triggered a lot even though there was a 10sec throttle
Probably because throttle also uses setTimout

kidroca committed:
Comment about `reachabilityRequestTimeout` and increase timeout a bit

kidroca commented:
Regarding issue 1 https://github.com/Expensify/App/pull/7825#pullrequestreview-891607801 and this thread https://github.com/Expensify/App/pull/7825#discussion_r812167915:
Here are the possible `isConnected` and `isInternetReachable` combinations:


```json5
// Connected to network with internet access
{"isInternetReachable":true,"isConnected":true,"type":"other","details":{"isConnectionExpensive":false}}

// Connected to network. Internet check running, previous internet check was `true` (same as above)
{"isInternetReachable":true,"isConnected":true,"type":"other","details":{"isConnectionExpensive":false}}

// Connected to network, internet check failed or took longer than 15 seconds (default timeout for NetInfo)
{"isInternetReachable":false,"isConnected":true,"type":"other","details":{"isConnectionExpensive":false}}

// Connected to network, internet check running, but previous internet check was `false`
{"isInternetReachable":null,"isConnected":true,"type":"other","details":{"isConnectionExpensive":false}}

// Not connected to network, internet check instantly fails
{"isInternetReachable":false,"isConnected":false,"type":"unknown","details":null}
```

The offline status is briefly displayed because `isInternetReachable` is `null`. 

There are 2 main ways to go about this:
1. Treat `isInternetReachable` being `null` as indication that we're offline (this is what we're currently doing)
2. Ony treat `isInternetReachable === false` as indication that we're offline. This would add a 15 sec. delay that we can perhaps try to reduce to 5 or 10 sec. 

I guess which of these we try to employ depends on the answer of the following question:
#### Do we want to indicate as soon as possible that App might be offline? 
- yes: then we try to make 1) work without initially showing an offline status
- no: go with 2) and maybe tweak the 15sec. timeout on `reachabilityRequestTimeout` in [the config](https://github.com/react-native-netinfo/react-native-netinfo#netinfoconfiguration)

I went with option 2 as it's simpler:

Now if an API request is taking 10 or more seconds a NetInfo check internet request is triggered
If the NetInfo request fails or also takes 10 or more seconds and event that we're offline is raised

kidroca commented:
> issue 2: The offline message comes and goes many times.

This is resolved by the submitted change using `isInternetReachable === false` 

> Issue 3: If you do above till 4, you will notice that wait time is very high.

This is expected for Web/Desktop
The added logic that triggers recheck is designed to address that, but yes if the user is not making requests there's nothing we can do but use more frequent polling. Do we want to poll for internet connectivity more frequent than once per minute?
As stated in NetInfo - internet reachability polling is only resorted to when the platform does not support this natively 

It's still better realizing that you're offline up to a minute after it happens than to never realize

kidroca commented:
✔️ Ready for review

parasharrajat commented:
Ok, I see. I will test it shortly.

parasharrajat commented:
Thanks for clarification and changes. Looks better now. 

> Issue 3: If you do above till 4, you will notice that wait time is very high.

Do we think we can resolve this by adding a global listener for NetInfo?
```js
 // whether a user has internet connectivity or not.
    unsubscribeFromNetInfo = NetInfo.addEventListener((state) => {
        Log.info('[NetworkConnection] NetInfo state', false, state);
        setOfflineStatus(!state.isInternetReachable);
 ```

kidroca committed:
Merge branch 'main' into kidroca/fix/connection-drop-is-not-detected

kidroca commented:
> > Issue 3: If you do above till 4, you will notice that wait time is very high.
> 
> Do you think we can improve this by adding a global listener for NetInfo?

How would that help?

parasharrajat commented:
Oh, So we always have an active listener Didn't notice it earlier. At this point, I say we are good.

parasharrajat reviewed (approved):
LGTM. 

cc: @pecanoro 

:ribbon: :eyes: :ribbon:  C+ reviewed

kidroca commented:
@pecanoro can you review this PR, it's been waiting for some time

pecanoro commented:
Sure, sorry for the delay, I was at a college career fair during the week so I wasn't on my computer.

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

kidroca commented:
Thank you 👍

marcaaron commented:
Just a heads up that this is causing CORS errors on dev where most internal engineers are using ngrok to tunnel to the local API. Is there any solution around this?

kidroca commented:
> Is there any solution around this?

Use a different URL here: https://github.com/kidroca/Expensify.cash/blob/b327db89745f1977dd40a5c68318389ee5d9d79b/src/libs/NetworkConnection.js#L60

Or disable the internet reachability test if ENV is DEV

kidroca commented:
Shouldn't `URL_API_ROOT` be the ngrok URL?

```js
const expensifyURLRoot = useNgrok && ngrokURL ? ngrokURL : expensifyComWithProxy;
...
URL_API_ROOT: expensifyURLRoot,
```

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/1965977435) to staging by @pecanoro in version: 1.1.42-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

marcaaron commented:
Throwing a deploy blocker on here temporarily while we try to figure out what is happening. I have a feeling these changes may be contributing to the instability brought up in [this thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1646838566936819)

marcaaron commented:
Ok, seems like it is only affecting dev 😄

roryabraham commented on src/libs/NetworkConnection.js:
```diff
@@ -86,6 +109,19 @@ function onReconnect(callback) {
+
+    Log.info('[NetworkConnection] recheck NetInfo');
+    hasPendingNetworkCheck = true;
+    unsubscribeFromNetInfo();
```
@kidroca Do you recall why you chose to unsubscribe and resubscribe to NetInfo rather than using [NetInfo.fetch](https://github.com/react-native-netinfo/react-native-netinfo#fetch) to recheck the network connection?

roryabraham replied:
Oh, looks like NetInfo will only trigger the recheck if you call `NetInfo.fetch` with a specific interface: https://github.com/react-native-netinfo/react-native-netinfo/blob/f05f7dd6e829135d907506e9270e729066b6e5d1/src/internal/state.ts#L100-L106

kidroca replied:
> Oh, looks like NetInfo will only trigger the recheck if you call `NetInfo.fetch` with a specific interface

Yes, there isn't an exposed way to force a recheck, it might not be necessary, though we decided to do it if any request is taking unusually long time (more than 10-15 sec)

What triggers the recheck is actually the call to `NetInfo.configure`

roryabraham replied:
> Yes, there isn't an exposed way to force a recheck

Yep, I created a PR for react-native-netinfo to add this and clean up this implementation a bit: https://github.com/react-native-netinfo/react-native-netinfo/pull/594
